examples/classification/accuracy_aware_worker.py

```diff
@@ -0,0 +1,192 @@
+
+        def autoq_eval_fn(model, eval_loader):
+            _, top5 = validate(eval_loader, model, criterion, config)
+            return top5
+
+        nncf_config = register_default_init_args(
+            nncf_config, init_loader, criterion, train_criterion_fn,
+            autoq_eval_fn, val_loader, config.device)
+
+    # create model
+    model = load_model(model_name,
+                       pretrained=pretrained,
+                       num_classes=config.get('num_classes', 1000),
+                       model_params=config.get('model_params'),
+                       weights_path=config.get('weights'))
+
+    model.to(config.device)
+
+    resuming_model_sd, _ = load_resuming_checkpoint(resuming_checkpoint_path)
+    compression_ctrl, model = create_compressed_model(model, nncf_config, resuming_state_dict=resuming_model_sd)
+
+    if config.to_onnx:
+        compression_ctrl.export_model(config.to_onnx)
+        logger.info("Saved to {}".format(config.to_onnx))
+        return
+
+    model, _ = prepare_model_for_execution(model, config)
+    if config.distributed:
+        compression_ctrl.distributed()
+
+    log_common_mlflow_params(config)
+
+    if config.execution_mode != ExecutionMode.CPU_ONLY:
+        cudnn.benchmark = True
+
+    if is_main_process():
+        print_statistics(compression_ctrl.statistics())
+
+    if config.mode.lower() == 'test':
+        validate(val_loader, model, criterion, config)
+
+    if config.mode.lower() == 'train':
+
+        # validation function that returns the target metric value
+        # pylint: disable=E1123
+        def validate_fn(model, epoch):
+            top1, _ = validate(val_loader, model, criterion, config, epoch=epoch)
+            return top1
+
+        # training function that trains the model for one epoch (full training dataset pass)
+        def train_epoch_fn(compression_ctrl, model, epoch, optimizer, lr_scheduler):
+            return train_epoch(train_loader, model, criterion, train_criterion_fn,
+                               optimizer, compression_ctrl, epoch, config)
+
+        # function that initializes optimizers & lr schedulers to start training
+        def configure_optimizers_fn():
+            params_to_optimize = get_parameter_groups(model, config)
+            optimizer, lr_scheduler = make_optimizer(params_to_optimize, config)
+            return optimizer, lr_scheduler
+
+        # register all these training-loop related funcs in nncf config
+        nncf_config = register_training_loop_args(nncf_config, train_epoch_fn, validate_fn,
+                                                  configure_optimizers_fn,
+                                                  tensorboard_writer=config.tb,
+                                                  log_dir=config.log_dir)
+
+        # run accuracy-aware training loop
+        model = run_accuracy_aware_compressed_training(model, compression_ctrl, nncf_config)
```

Comment:
My opinion is that the accuracy-aware algorithm does not fit in the general pipeline for the compression algorithms by design. Thus, we can design API for this algorithm without taking into account abstractions that are used in the compression pipeline. I think we should pay more attention to making the algorithm API as user-friendly as possible. I mean, we should think about simplifying API and how we will document itself. 

One possibility for simplification is not to use NNCFConfig and explicitly define the parameters. Another possibility is to provide a class instead of a function. This allows you to separate the parameters of the algorithm and the required parameters to run the algorithm. 

I have several suggestions:

```
estimator = AccuracyAwareEstimator(nncf_config)

estimator.run(
    model=model,
    compression_controller=compression_ctrl,
    train_epoch_fn=train_epoch_fn,
    validate_fn=validate_fn,
    configure_optimizers_fn=configure_optimizers_fn,
    tensorboard_writer=config.tb,
    )
```
```
config = AccuracyAwareConfig(
            is_higher_metric_better=True,
            initial_compression_rate_step=0.1,
            compression_rate_step_reduction_factor=0.5,
            minimal_compression_rate_step=0.025,
            patience_epochs=10,
            maximal_total_epochs=float('inf'),
)

config = nncf_config.get_accuracy_aware_config()

model = run_accuracy_aware_compressed_training(
    config=config,
    model=model,
    compression_controller=compression_ctrl,
    train_epoch_fn=train_epoch_fn,
    validate_fn=validate_fn,
    configure_optimizers_fn=configure_optimizers_fn)
```

Reply:
Right now it's implemented through a `TrainingLoop` object (similar to `AccuracyAwareEstimator` above). The scope covered here is implemented via `AdaptiveCompressionTrainingLoop` (i.e. compression rate is varied during training). We would implement a simpler training loop scheme with a constant target compression rate (e.g. a `EarlyExitCompressionTrainingLoop`), which would terminate once an accuracy drop criterion is satisfied, as per our discussion with @AlexKoff88